Initial steps towards remoting compression

ktoso opened the pull request:
Will resolve https://github.com/akka/akka/issues/20455

Work In Progress, in case someone wants to comment early on.
Mostly "on the side" from the actual remoting for now

Concept diagram of incoming / emitting compression / applying compression:

![artery-diagram-simple](https://cloud.githubusercontent.com/assets/120979/15773662/7c3eca00-2978-11e6-9446-78e59743b374.jpg)

ktoso committed:
+art #20455 HeavyHitters and CountMinSketch prepared for Compression

ktoso committed:
+art #20455 WIP compression tables and integration specs

patriknw commented on akka-actor/src/main/scala/akka/routing/MurmurHash.scala:
```diff
@@ -127,8 +127,8 @@ object MurmurHash {
-   *  where the order of appearance of elements does not matter.
-   *  This is useful for hashing sets, for example.
+   * where the order of appearance of elements does not matter.
+   * This is useful for hashing sets, for example.
```
revert those formatting changes to avoid merge problems later

ktoso replied:
reverted

patriknw commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,338 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
this is not our ordinary header, it was copied from clearspring but since it contains no copyright information, only ordinary license info (which is the same as ours) I see no reason to include it here. The later comment about derived work from clearspring should be enough

ktoso replied:
Okey, amending

patriknw commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,338 @@
+   * Andrzej Bialecki (ab at getopt org).
+   * </p>
+   */
+  private static class MurmurHash {
```
how is this different from `akka.routing.MurmurHash` or Scala's `MurmurHash3`? could we use any of those instead?

ktoso replied:
I think there's a TODO around here to clean them up (have just one), yes :)

ktoso replied:
(took this one for ease of integrating the CMS impl, will need to adjust APIs in ours – will do this week)

drewhk replied:
I know something better: **Use CRC32!** First, it is a reusable mutable construct which takes an array instead of allocating one, second, since Java8 **it is a compiler intrinsic** compiling down to CPU supported CRC instructions! Tops, CRC32 is still a bloody good hash function. You can try with turning the intrinsics off (-XX:-UseCLMUL -XX:-UseCRC32Intrinsics) and on (I think that's the default).

drewhk replied:
Of course benchmark it :)

drewhk replied:
I did some benchmark of hashes:

```
[info] Benchmark               (lengthOfArray)   Mode  Cnt   Score   Error   Units
[info] HashesBench.adler32                 100  thrpt    6  14,354 ? 2,779  ops/us
[info] HashesBench.adler32                 200  thrpt    6   9,207 ? 0,312  ops/us
[info] HashesBench.crc32                   100  thrpt    6  11,483 ? 6,097  ops/us
[info] HashesBench.crc32                   200  thrpt    6   9,686 ? 0,190  ops/us
[info] HashesBench.fnv32                   100  thrpt    6   9,871 ? 0,011  ops/us
[info] HashesBench.fnv32                   200  thrpt    6   4,533 ? 0,006  ops/us
[info] HashesBench.fnv64                   100  thrpt    6   9,838 ? 0,033  ops/us
[info] HashesBench.fnv64                   200  thrpt    6   4,519 ? 0,018  ops/us
[info] HashesBench.murmur3                 100  thrpt    6  13,745 ? 0,114  ops/us
[info] HashesBench.murmur3                 200  thrpt    6   7,707 ? 1,650  ops/us
[info] HashesBench.murmurAkka              100  thrpt    6   3,531 ? 0,044  ops/us
[info] HashesBench.murmurAkka              200  thrpt    6   1,570 ? 0,237  ops/us
```

Previous results were rather wonky (some hashes said 800k ops/us wat) so I stored the result in a volatile var in the benchmark. These results look better. Murmur3 is `scala.util.hashing.MurmurHash3`, FNV are implementations from the internet, murmurAkka is the one that is included in akka itself. CRC32 and Adler32 are the JDK provided ones.

drewhk replied:
CRC vs Murmur3

```
[info] Benchmark            (lengthOfArray)   Mode  Cnt   Score   Error   Units
[info] HashesBench.crc32                 10  thrpt  100  36,812 ? 1,316  ops/us
[info] HashesBench.crc32                 50  thrpt  100  16,378 ? 1,377  ops/us
[info] HashesBench.crc32                100  thrpt  100  12,185 ? 0,611  ops/us
[info] HashesBench.crc32                200  thrpt  100   9,859 ? 0,149  ops/us
[info] HashesBench.murmur3               10  thrpt  100  68,119 ? 0,354  ops/us
[info] HashesBench.murmur3               50  thrpt  100  24,207 ? 0,480  ops/us
[info] HashesBench.murmur3              100  thrpt  100  13,764 ? 0,014  ops/us
[info] HashesBench.murmur3              200  thrpt  100   7,195 ? 0,003  ops/us
```

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/CompressionProtocol.scala:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
```
2016 would be enough

ktoso replied:
:+1:

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/CompressionProtocol.scala:
```diff
@@ -0,0 +1,18 @@
+  // TODO do we bind into ActorRef or keep it "could be anything"?
+
+  /** Sent by the "receiving" node after allocating a compression id to a given [[akka.actor.ActorRef]] */
+  final case class ActorRefCompressionAdvertisement(ref: ActorRef, id: Long)
```
will the ids be Long, I thought the table would be small?

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/CompressionProtocol.scala:
```diff
@@ -0,0 +1,18 @@
+
+object CompressionProtocol {
+
+  // TODO do we bind into ActorRef or keep it "could be anything"?
```
yes, it will be String

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,88 @@
+import scala.collection.immutable
+
+// TODO describe / think about how we could handle an actor dying and removing from "top"
+// TODO this is the scenario when "super traffic actor" goes away and we want to not use up the slot in HH
```
doesn't that fix itself after some time?

drewhk replied:
I imagined that we roughly use the following:
 1. Use table A, and in the meantime build new table B (from scratch!)
 2. Eventually start advertising B
 3. We see remote using B, switch to B and start from 1. (now B becoming A)

drewhk replied:
And that automatically removes no longer referenced actors. Also, this works in the weird case if an actor is dead here, but someone still keeps sending tons of remote messsages to it. Even though those messages are dropped *here* they still travel through the wire so it still makes sense to compress them.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,88 @@
+ * for a discussion about the assumptions made and guarantees about the Heavy Hitters made in this model.
+ * We assume the Cash Register model in which there are only additions, which simplifies HH detecion significantly.
+ */
+sealed class HeavyHitters[T](strategy: HeavyHittersStrategy[T]) {
```
make it internal api right away, so we don't forget

ktoso replied:
done

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,88 @@
+  // TODO technically could be optimised depending on strategy (priority queue?) (store in order of counts, drop last if exceeded size)
+  private[this] var _items = List.empty[HeavyHitter[T]]
+
+  def items: immutable.Set[T] = _items.map(_.item).toSet // TODO avoid the toSet? does anyone care?
```
I don't know where it's used yet, but `breakOut` is better

ktoso replied:
thanks - hope to get rid of this somehow btw

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala:
```diff
@@ -0,0 +1,100 @@
+  // TODO calibrate properly
+  private[this] val cms = new CountMinSketch(10, 10, 0)
+
+  private[this] var compressionAllocations = Map.empty[Long, String] // TODO replace with a specialized LongMap
```
how many are we talking about? could perhaps be an `Array`?

ktoso replied:
Yes, all these datastructures will be optimised for the Long/Int keys

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala:
```diff
@@ -0,0 +1,100 @@
+ * Mutable and not thread-safe.
+ *
+ * Compression flow goes like:
+ * [1]  on each message we add the actor path here
```
Will we have one compression table per stream or one for all streams of the destination node?
If it is the latter it might be problematic with thread safety, since this will be updated by multiple threads (stream stages).
How costly is `addAndEstimateCount`? Would it be worth doing the updates async in the background (e.g. by an actor), but we also want to keep allocations low?
We could also use sampling if too costly to do on each message.

ktoso replied:
> Will we have one compression table per stream or one for all streams of the destination node?
If it is the latter it might be problematic with thread safety, since this will be updated by multiple threads (stream stages).

One per stream, so tables assume access is properly synchronised and don't guard against concurrent access.

> How costly is addAndEstimateCount? 

CMS adding is nearly O(1). It should be relatively fast, it is "hash the string, add it into m slots" (m relates to prealocated size of table). I'll measure it in more detail ofc.

> Would it be worth doing the updates async in the background (e.g. by an actor), but we also want to keep allocations low? 
Maybe doing the allocation could be moved over, will think about it definitely.

> We could also use sampling if too costly to do on each message.

I don't think we'll have to, the datastructure is aimed exacly at cases that we have, high throughput counting.

drewhk replied:
We should try to keep caching the hash value where it is possible.

patriknw commented:
looking good

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Codecs.scala:
```diff
@@ -29,23 +29,13 @@ class Encoder(
-        }
-        headerBuilder.recipientActorRef = recipientStr
+        // internally compression is applied by the builder:
+        headerBuilder.recipientActorRef = send.recipient.path.toSerializationFormat
```
this is also something we must optimize, `toSerializationFormat` is way to costly, perhaps also something we can cache in the RemoteActorRef

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala.orig:
```diff
@@ -0,0 +1,603 @@
+/**
```
it would be nice if you could add `*.orig` to .gitignore

ktoso replied:
yes, done now. Sorry for that

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -498,72 +485,93 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
+    val advertActorRef = (ref: String, id: Int) ⇒ {
+      val addr = ActorPath.fromString(ref).address // TODO expensive      
+      log.error(s"advertise compression to ${addr}, of [$ref] => [$id] ")
+      sendControl(addr, CompressionProtocol.ActorRefCompressionAdvertisement(addr, ref, id))
```
How does the versioning work? Is that the `id`? I think it's also important to include the actor system `UID`, i.e. use `UniqueAddress`, so that we never ever use a table that doesn't match the incarnation

ktoso replied:
doesn't have versions yet, like I said in standup yesterday - currently assumes "there is the table", no swapping it for a new one

ktoso replied:
The ID here is what ID we allocated in our table for this actor

ktoso commented:
I added a concept diagram to the first post btw.

ktoso commented:
Rebased, squashed and passing tests.
It currently slows down the throughput - as expected actually, due to all the heavy structures – will now work towards optimising and cleaning up things. I'll also prepare a TODO list for next steps.

I'm keeping the `COMPRESSEDSpec` for now as it allows simple side-by-side comparing of compressed run vs. normal run.

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,347 @@
+  private void initTablesWith(int depth, int width, int seed) {
+    this.table = new long[depth][width];
+    this.hashA = new long[depth];
+    Random r = new Random(seed);
```
use threadlocalrandom?

ktoso replied:
good idea

ktoso replied:
not hopping to that yet (it would have to drop the seed), can do so later

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,347 @@
+    this.hashA = new long[depth];
+    Random r = new Random(seed);
+    // We're using a linear hash functions
+    // of the form (a*x+b) mod p.
```
Hm, maybe better to use the FNV hash (http://www.isthe.com/chongo/tech/comp/fnv/). That is just a simple (p*x ^ byteToHash) operation, no need for costly modulo. You can simply get a family of them by using different initial seeds.

drewhk replied:
Ah, I see, it uses a nice trick to eliminate the modulus. Still, if this ever becomes a bottleneck an FNV hash can be a good alternative.

ktoso replied:
(the actual paper is http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf )

drewhk replied:
> I'm not sure about changing the hash used, the alg heavily rides on hash properties/distribution

I am not sure, it uses Murmur to calculates buckets in one case, and in other case it calculates it using this prime modulo hash, so I am not convinced it actually relies on the hash function that much. Anyway, probably won't be the slowest thing here.

ktoso replied:
For reference, we'll stay with murmur3

drewhk replied:
The one included here is Murmur3?

ktoso replied:
This is murmur2. I did not manage to move to Scalas murmur3 yet due to java
scala interop issues. Its on the list to do.

I'll write up.remaining things to be done like that. I first need to get it
working then replace hashes etc
On Jun 7, 2016 10:27, "drewhk" <notifications@github.com> wrote:

> In
> akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java
> <https://github.com/akka/akka/pull/20546#discussion_r66028767>:
>
> > +  CountMinSketch(int depth, int width, int size, long[] hashA, long[][] table) {
> > +    this.depth = depth;
> > +    this.width = width;
> > +    this.eps = 2.0 / width;
> > +    this.confidence = 1 - 1 / Math.pow(2, depth);
> > +    this.hashA = hashA;
> > +    this.table = table;
> > +    this.size = size;
> > +  }
> > +
> > +  private void initTablesWith(int depth, int width, int seed) {
> > +    this.table = new long[depth][width];
> > +    this.hashA = new long[depth];
> > +    Random r = new Random(seed);
> > +    // We're using a linear hash functions
> > +    // of the form (a*x+b) mod p.
>
> The one included here is Murmur3?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66028767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAHYk1vSvMnaLhe8iC2S0xmZTbjMvWiJks5qJSuIgaJpZM4IhRoS>
> .
>

drewhk replied:
Just copy out the necessary code from Murmur3 impl?

On Tue, Jun 7, 2016 at 10:57 AM, Konrad Malawski <notifications@github.com>
wrote:

> In
> akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java
> <https://github.com/akka/akka/pull/20546#discussion_r66033001>:
>
> > +  CountMinSketch(int depth, int width, int size, long[] hashA, long[][] table) {
> > +    this.depth = depth;
> > +    this.width = width;
> > +    this.eps = 2.0 / width;
> > +    this.confidence = 1 - 1 / Math.pow(2, depth);
> > +    this.hashA = hashA;
> > +    this.table = table;
> > +    this.size = size;
> > +  }
> > +
> > +  private void initTablesWith(int depth, int width, int seed) {
> > +    this.table = new long[depth][width];
> > +    this.hashA = new long[depth];
> > +    Random r = new Random(seed);
> > +    // We're using a linear hash functions
> > +    // of the form (a*x+b) mod p.
>
> This is murmur2. I did not manage to move to Scalas murmur3 yet due to
> java scala interop issues. Its on the list to do. I'll write up.remaining
> things to be done like that. I first need to get it working then replace
> hashes etc
> … <#m_1377374578521158062_>
> On Jun 7, 2016 10:27, "drewhk" ***@***.***> wrote: In
> akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java <#20546
> (comment) <https://github.com/akka/akka/pull/20546#discussion_r66028767>>:
> > + CountMinSketch(int depth, int width, int size, long[] hashA, long[][]
> table) { > + this.depth = depth; > + this.width = width; > + this.eps = 2.0
> / width; > + this.confidence = 1 - 1 / Math.pow(2, depth); > + this.hashA =
> hashA; > + this.table = table; > + this.size = size; > + } > + > + private
> void initTablesWith(int depth, int width, int seed) { > + this.table = new
> long[depth][width]; > + this.hashA = new long[depth]; > + Random r = new
> Random(seed); > + // We're using a linear hash functions > + // of the form
> (a*x+b) mod p. The one included here is Murmur3? — You are receiving this
> because you authored the thread. Reply to this email directly, view it on
> GitHub <
> https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66028767>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/AAHYk1vSvMnaLhe8iC2S0xmZTbjMvWiJks5qJSuIgaJpZM4IhRoS>
> .
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66033001>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA45vfOLs8lTf5c8xLQjCpAmRKirudOrks5qJTJlgaJpZM4IhRoS>
> .
>

ktoso replied:
Yeah, sounds good as well - will do that.

-- 
Konrad `ktoso` Malawski
Akka <http://akka.io> @ Lightbend <http://lightbend.com>

On 7 June 2016 at 12:58:21, drewhk (notifications@github.com) wrote:

In akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java
<https://github.com/akka/akka/pull/20546#discussion_r66049464>:

> +  CountMinSketch(int depth, int width, int size, long[] hashA, long[][] table) {
> +    this.depth = depth;
> +    this.width = width;
> +    this.eps = 2.0 / width;
> +    this.confidence = 1 - 1 / Math.pow(2, depth);
> +    this.hashA = hashA;
> +    this.table = table;
> +    this.size = size;
> +  }
> +
> +  private void initTablesWith(int depth, int width, int seed) {
> +    this.table = new long[depth][width];
> +    this.hashA = new long[depth];
> +    Random r = new Random(seed);
> +    // We're using a linear hash functions
> +    // of the form (a*x+b) mod p.

Just copy out the necessary code from Murmur3 impl? On Tue, Jun 7, 2016 at
10:57 AM, Konrad Malawski <notifications@github.com> wrote:
… <#>
In
akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java
<#20546
(comment) <https://github.com/akka/akka/pull/20546#discussion_r66033001>>:
> + CountMinSketch(int depth, int width, int size, long[] hashA, long[][]
table) { > + this.depth = depth; > + this.width = width; > + this.eps = 2.0
/ width; > + this.confidence = 1 - 1 / Math.pow(2, depth); > + this.hashA =
hashA; > + this.table = table; > + this.size = size; > + } > + > + private
void initTablesWith(int depth, int width, int seed) { > + this.table = new
long[depth][width]; > + this.hashA = new long[depth]; > + Random r = new
Random(seed); > + // We're using a linear hash functions > + // of the form
(a*x+b) mod p. This is murmur2. I did not manage to move to Scalas murmur3
yet due to java scala interop issues. Its on the list to do. I'll write
up.remaining things to be done like that. I first need to get it working
then replace hashes etc … <#m_1377374578521158062_> On Jun 7, 2016 10:27,
"drewhk" ***@***.***> wrote: In
akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java <
#20546 <https://github.com/akka/akka/pull/20546> (comment) <#20546 (comment)
<https://github.com/akka/akka/pull/20546#discussion_r66028767>>>: > +
CountMinSketch(int depth, int width, int size, long[] hashA, long[][]
table) { > + this.depth = depth; > + this.width = width; > + this.eps = 2.0
/ width; > + this.confidence = 1 - 1 / Math.pow(2, depth); > + this.hashA =
hashA; > + this.table = table; > + this.size = size; > + } > + > + private
void initTablesWith(int depth, int width, int seed) { > + this.table = new
long[depth][width]; > + this.hashA = new long[depth]; > + Random r = new
Random(seed); > + // We're using a linear hash functions > + // of the form
(a*x+b) mod p. The one included here is Murmur3? — You are receiving this
because you authored the thread. Reply to this email directly, view it on
GitHub <
https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66028767>,
or mute the thread <
https://github.com/notifications/unsubscribe/AAHYk1vSvMnaLhe8iC2S0xmZTbjMvWiJks5qJSuIgaJpZM4IhRoS>
. — You are receiving this because you commented. Reply to this email
directly, view it on GitHub <
https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66033001>,
or mute the thread <
https://github.com/notifications/unsubscribe/AA45vfOLs8lTf5c8xLQjCpAmRKirudOrks5qJTJlgaJpZM4IhRoS>
.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/akka/akka/pull/20546/files/1bc325ec93fe3f56dc850ad5a449f3f7d9115c60#r66049464>,
or mute the thread
<https://github.com/notifications/unsubscribe/AAHYk8yM10KESrQp0ABP77I5SlRPFW6Zks5qJU7NgaJpZM4IhRoS>
.

ktoso replied:
Note to self, not done yet.

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,347 @@
+      // parameter to constructor.
+      throw new IllegalArgumentException("Negative increments not implemented");
+    }
+    final int[] buckets = MurmurHash.hashBuckets(item, depth, width);
```
allocates an array all the time. Would be better to take an array as mutable parameter in `hashBuckets` and then have a reused array in the class.

ktoso replied:
good idea, also I think we'd be able to recycle this one as we only ever add one item at a time

patriknw replied:
capture such ideas as TODO or FIXME

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,347 @@
+   * with probability {@code confidence}.
+   */
+  public long estimateCount(String item) {
+    int[] buckets = MurmurHash.hashBuckets(item, depth, width);
```
same, here, reuse array.

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -498,72 +487,94 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
   }
 
   def inboundLargeFlow: Flow[EnvelopeBuffer, ByteString, NotUsed] = {
+    // TODO large flow does not care about compression IMO. right?  
```
Probably not, although it does not hurt to cache ActorRef instances still, some optimizations might rely on that.

patriknw replied:
and small messages may still be sent over the large stream

ktoso replied:
Good point @patriknw.
So if we'd share the same table in this and the other flow then it would have to guard from concurrent access I think... Will think about solving this (2 tables also introduces some questions)

patriknw replied:
I thought we would have one table per stream, for the sake of no concurrency.
There must anyway be a table for the control stream. E.g. heartbeat messages will be sent over that one.

drewhk replied:
Why share? There can't be any common traffic between the two. A ref is either routed to large messages channel or to the other, there is no intersection.

ktoso replied:
Hm, right - so 3 tables. Need to add some demultiplexing on the CompressionAdvertisement – will add an int to identify for which channel the compression is meant.

patriknw replied:
then use the same stream ids as Aeron, see `controlStreamId` in `ArteryTransport` (don't ask me why there is no 2)

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -125,8 +130,10 @@ private[akka] class Association(
   // OutboundContext
-  override def sendControl(message: ControlMessage): Unit =
+  override def sendControl(message: ControlMessage): Unit = {
+    println("sendControl === " + message)
```
println

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -283,7 +290,19 @@ private[akka] class Association(
+  // FIXME: Compression table must be owned by each channel instead of having a global one
+  private def createCompressionTable(): LiteralCompression =
+    if (transport.provider.remoteSettings.ArteryCompressionSettings.enabled) {
+      println(s"createCompressionTable for $this")
```
println and also in other places

ktoso replied:
right, please ignore the println and log.error's - still some raw debugging output in this PR

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/NoLiteralCompression.scala:
```diff
@@ -0,0 +1,24 @@
+  override def compressActorRef(ref: String): Int = -1
+
+  override def hitActorRef(ref: String): Unit = ()
+  override def decompressActorRef(idx: Int): String = ???
```
throw `UnsupportedOpEx`

ktoso replied:
Done in upcoming push

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala:
```diff
@@ -69,18 +69,34 @@ private[remote] object EnvelopeBuffer {
+private[remote] trait InboundCompression {
+  /** Incoming side counts how many times an ActorRef is used and then issues compression if it is considered "heavy" */
+  def hitActorRef(ref: String): Unit
   def decompressActorRef(idx: Int): String
```
This was String only for my convenience, but in reality this must be an `ActorRef`! Otherwise you will need to allocate a new one every time which is one of the things we want to exactly avoid.

ktoso replied:
Duh, exactly what I was thinking but then fit into the existing interfaces.
All tables are ready to hold anything – will hold ActorRefs 👍

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/ActualCompressionTables.scala:
```diff
@@ -0,0 +1,61 @@
+import akka.remote.artery.LiteralCompression
+
+/** INTERNAL API */
+object ActualCompressionTables {
```
er, naming :)

ktoso replied:
This class will go away in the form it exists right now I think, likely better to split up into incoming/outgoing (note that internally it is anyway already).

ktoso replied:
Yes, agree on rename of course :)

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/CompressionSettings.scala:
```diff
@@ -0,0 +1,47 @@
+package akka.remote.artery.compress
```
copyright

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,107 @@
+ * INTERNAL API
+ * [[HeavyHitters]] that never triggers a detection.
+ */
+private[remote] object NeverHeavyHitters extends HeavyHitters[Any](null) {
```
Do we need all these strategies? Isn't it simpler to take one (topN makes most sense to me as it gives a fixed size table) and optimize the hell out of it?

patriknw replied:
I agree, we want fixed size anyway so we can use arrays.

ktoso replied:
I agree on making an explicit choice about it, think "topN" is the one we'll want?
I wasn't sure which strategy would make sense so for initial phase hoped to benchmark around (and ask people for stats)

drewhk replied:
There is no difference between topPercentage or topN apart from that topN keeps the table bounded while percentage does not. Since we want to keep the table advertisements bounded, topN is the natural choice I think.

ktoso replied:
Ok, TopN in Array and optimising the hell of it makes sense (the current impl dominates everything by the iterating over the List all the time)

drewhk replied:
See my other remarks for hints. Iterating the list is super slow because of thrashing the cache, takes at least 10x times longer to access something not on cache line.

ktoso replied:
I know, meh makes me want to not share WIP until done optimising and feature complete :<

ktoso replied:
Anyway, please do keep those comments commit :-) Found a few ideas I didn't think of already! Thanks a lot :)

drewhk replied:
> I know, meh makes me want to not share WIP until done optimising and feature complete :<

The whole point of sharing early is so you can get nice ideas that you didn't think of. It is much easier to give ideas on improving something that *already exists* than to just pull them out of someone's head. I.e. sharing early gives others a better chance to provide ideas as they are already inspired by the thing already there.

ktoso replied:
I agree, please do keep commenting. I'll keep sharing early :-)

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,107 @@
+ */
+private[remote] sealed class HeavyHitters[T](strategy: HeavyHittersStrategy[T]) {
+  // TODO technically could be optimised depending on strategy (priority queue?) (store in order of counts, drop last if exceeded size)
+  private[this] var _items = List.empty[HeavyHitter[T]]
```
Make this a priority queue (java.util.PriorityQueue is a binary heap, and hence array backed). Traversing lists is super expensive, everything is a cache miss basically.

drewhk replied:
Of course custom binary heap is not hard to write, and might provide even more performance, but I am fine with the Java one for now.

ktoso replied:
I'll move to the PQ for now, removed all the other strategies along the way, will measure

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,107 @@
+  // TODO technically could be optimised depending on strategy (priority queue?) (store in order of counts, drop last if exceeded size)
+  private[this] var _items = List.empty[HeavyHitter[T]]
+
+  def items: immutable.Set[T] = _items.map(_.item)(scala.collection.breakOut).toSet // TODO avoid the toSet? does anyone care?
```
give back an `Iterator` to the backing PriorityQueue

ktoso replied:
definitely! done

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/HeavyHitters.scala:
```diff
@@ -0,0 +1,107 @@
+   * @return `true` if the added item is indeed a heavy hitter
+   */
+  def update(item: T, count: Long): Boolean =
+    _items.find(_.item == item) match { // TODO optimise? could it be the same case always?
```
if using a PriorityQueue, you need a linear traversal here, but since it is Array backed it should be cache friendly. We can do even better with a custom queue if:
 - maintain a queue built from two separate arrays, one containing item hash, the other the actual item themself
 - search in the first queue by hash value (no pointer chasing since it is an Array[Long]) and check the second queue for equality if a matching hash is found
 - of course the two queues must be kept in sync, I.e. it is *one* priority queue just blasted into multiple arrays for cache friendliness (since we have no record types)

drewhk replied:
Also, linear scanning for Long's (hashes) will benefit from streaming memory access, so cache prefetches will be also on our side.

drewhk replied:
Actually, can be a third Array as well, tracking the actual costs, so binary heap rotations don't need to touch the type `T` at all, only the reference to it.

drewhk replied:
Thinking about it even more, there is probably not much sense in a priority queue either, since removal of the smallest will be less frequent than updating values. Sorted arrays hence make more sense, since then `find` is just a  binary search. Removal then will make a hole, which you need to fix by `System.arrayCopy`. That should be also fast.

ktoso replied:
removal happens a lot, since we want to keep N in memory, so we add (so we're at N+1 in collection), then remove "lightest" to keep the the N heaviest. This is trashing a lot though, going towards the arrays now – will definitely be much better, pre allocated + longs + no envelope <3

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,347 @@
+    public static int[] hashBuckets(String key, int hashCount, int max) {
+      byte[] b;
+      try {
+        b = key.getBytes("UTF-16");
```
Use the `Unsafe` trick @patriknw used to access the backing array directly

drewhk commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala:
no need for this val

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/TopHeavyHittersScanning.java:
```diff
@@ -0,0 +1,82 @@
+ */
+package akka.remote.artery.compress;
+
+public class TopHeavyHittersScanning {
```
why Java?

ktoso replied:
Silly reason, was easier to write this: https://github.com/akka/akka/pull/20546/files/9c5ebdbf283862da0d042fabf0c745aec3b0e2d0..11feca1d6b2b2e587397286ef85d457d0eea939b#diff-84014b2f094e6f03567127be21cfc2b8R54
in Scala I'll to for-loop and put a weird index to terminate loop, can do so though if we want to be pure scala here (technically this wastes some bytes of space, but doesnt matter in face of the large arrays inside anyway(

drewhk replied:
In scala I would use a while loop

ktoso replied:
yeah, that's what I'm saying while + force its termination via putting a high index.
Anyway, will rewrite to pure scala :)

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/TopHeavyHittersScanning.java:
```diff
@@ -0,0 +1,82 @@
+  private int _lowestHitterIdx = 0;
+  
+  protected final int[] hashes;
+  protected final Object[] items;
```
why Object?

ktoso replied:
Was bezing lazy; see "generic array creation"; I'll fix though, should be T

ktoso replied:
Moved most things into Scala, the array scanning remains in Java (see benchmark)

drewhk commented on akka-bench-jmh/src/main/scala/akka/remote/compress/HeavyHittersBenchmark.scala:
```diff
@@ -3,55 +3,64 @@
 
 import akka.remote.artery.compress.TopHeavyHitters
 import org.openjdk.jmh.annotations._
+import org.openjdk.jmh.infra.Blackhole
```
a-ha, this is useful to know!

ktoso replied:
Right, one should either return the result of the benchmark (then JMH does the blackhole), or if it's not possible for some reason blackhole yourself.
It mostly is about a volatile write but not only, it does some more magic to disalow the jit to optimise stuff away there

drewhk commented on akka-remote/src/main/java/akka/remote/artery/compress/TopHeavyHittersScanning.java:
```diff
@@ -0,0 +1,82 @@
+    items = new Object[max];
+  }
+
+  protected boolean insertIfHeavy(Object item, int hashCode, long count) {
```
why not expose any methods as public? I like to avoid inheritance. Or if you want to force inheritance for some reason, then I would prefer making these `protected final` and the fields it touches to be `private`.

ktoso commented on akka-bench-jmh/src/main/scala/akka/remote/compress/HeavyHittersBenchmark.scala:
```diff
@@ -3,55 +3,64 @@
+ * [info] Benchmark                                  (n)   Mode  Cnt          Score         Error  Units
+ * [info] HeavyHittersBenchmark.updateHitter        8192  thrpt   40     357405.512 ±    3329.008  ops/s
+ * [info] HeavyHittersBenchmark.updateNotHitter     8192  thrpt   40  259032711.743 ± 7199514.142  ops/s
+ * [info] HeavyHittersBenchmark.updateRandomHitter  8192  thrpt   40    2105102.088 ±   18214.624  ops/s
```
Though I think we'll end up maintaining all these tables asynchronously from the incoming messages stream in any case - we'll see soon

ktoso commented:
For reference why the scanning of arrays I decided to keep in Java:

Scala Scanning

```scala
private def findHashIdx(searchFromIndex: Int, hashCode: HashCodeVal): Int = { 
  // var it = -1 this is even worse
   var i = searchFromIndex
   while (i < hashes.length/* && it == -1*/) { 
    if (hashes(i) == hashCode.get) return i
     i += 1
   }
   -1
 }
Results

```
[info] Benchmark                                  (n)   Mode  Cnt          Score          Error  Units
[info] HeavyHittersBenchmark.updateHitter        8192  thrpt   40     366256.293 ±    11145.098  ops/s
[info] HeavyHittersBenchmark.updateNotHitter     8192  thrpt   40  262319483.012 ± 17092723.289  ops/s
[info] HeavyHittersBenchmark.updateRandomHitter  8192  thrpt   40    1600544.397 ±    19960.740  ops/s
```

Java Scanning

```java
public static int findHashIdx(int[] hashes, int searchFromIndex, int hashCode) { 
for (int i = searchFromIndex; i < hashes.length; i++)
     if (hashes[i] == hashCode) return i;
   return -1; 
}
``` 

Results

```
[info] Benchmark (n) Mode Cnt Score Error Units
[info] HeavyHittersBenchmark.updateHitter 8192 thrpt 40 430302.379 ± 28669.315 ops/s
[info] HeavyHittersBenchmark.updateNotHitter 8192 thrpt 40 262163962.484 ± 8643343.025 ops/s
[info] HeavyHittersBenchmark.updateRandomHitter 8192 thrpt 40 2077940.873 ± 45405.661 ops/s
```

ktoso commented:
Seems I didn't push last nights work, will do so shortly.

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/Codecs.scala:
```diff
@@ -50,20 +45,14 @@ class Encoder(
-            //            }
-            headerBuilder.senderActorRef = sender // TODO cache not used now
+            // FIXME: Since we must include the loacl address, and API works on actor refs 
+            headerBuilder.senderActorRef = new MinimalActorRef {
```
This was previously string based, thus it was `sender.path.toSerializationFormatWithAddress(localAddress)`.
Currently builder works on ActorRef, as it may store it in the table, so this workaround is in place...

I'll rethink when to require passing around ActorRef (into builder / tables) and when string repr is enough.

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala:
```diff
@@ -142,7 +143,7 @@ private[remote] final class HeaderBuilderImpl(val compression: LiteralCompressio
   def senderActorRef_=(ref: ActorRef): Unit = {
     _senderActorRefIdx = compression.compressActorRef(ref)
-    if (_senderActorRefIdx == -1) _senderActorRef = ref.path.toSerializationFormat
+    if (_senderActorRefIdx == -1) _senderActorRef = Serialization.serializedActorPath(ref) // includes local address from `currentTransportInformation`
```
`toSerializationFormat` does not include the address from the `currentTransportInformation`, one has to use serializedActorPath

patriknw commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputCOMPRESSEDSpec.scala:
```diff
@@ -0,0 +1,284 @@
+import com.typesafe.config.ConfigFactory
+import java.net.InetAddress
+
+object MaxThroughputCOMPRESSEDSpec extends MultiNodeConfig {
```
before merging this should be removed, right?

ktoso replied:
Yes, I'll remove now since compression is now usable without breaking anything else.

patriknw commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala:
```diff
@@ -36,15 +38,14 @@ object MaxThroughputSpec extends MultiNodeConfig {
          }
        }
-       """)))
+       """.stripMargin)))
```
?

ktoso replied:
reverted changes to this file

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -124,6 +124,21 @@ akka {
+          enabled = off
+          max = 256
+        }
+      }
```
should probably go into the `advanced` section, as most things

patriknw commented on akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala:
```diff
@@ -362,13 +362,15 @@ private[akka] class RemoteActorRefProvider(
+        if (hasAddress(address))
+          local.resolveActorRef(rootGuardian, elems)
         else try {
+          // TODO this fails since we get an akka:// path address here, since the testActor is "akka://"
```
?

patriknw commented on akka-remote/src/main/scala/akka/remote/RemoteActorRefProvider.scala:
```diff
@@ -362,13 +362,15 @@ private[akka] class RemoteActorRefProvider(
         } catch {
           case NonFatal(e) ⇒
-            log.warning("Error while resolving ActorRef [{}] due to [{}]", path, e.getMessage)
+            log.error(e, "Error while resolving ActorRef [{}] with local address [{}] due to [{}]", path, localAddress, e.getMessage)
```
change back?

ktoso replied:
👍

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -388,35 +378,24 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
+    val compression = createInboundCompressionTable()
+
+    runInboundControlStream(compression)
+    runInboundOrdinaryMessagesStream(compression)
```
if we start without versioning and identifiers we should only use compression for the OrdinaryMessagesStream, i.e. one table

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/CompressionProtocol.scala:
```diff
@@ -0,0 +1,37 @@
+   * INTERNAL API
+   * Sent by the "receiving" node after allocating a compression id to a given [[akka.actor.ActorRef]]
+   */
+  private[remote] final case class ActorRefCompressionAdvertisement(from: Address, ref: ActorRef, id: Int)
```
`from` should be `UniqueAddress`
ah, so currently each actor ref is advertised separately, I thought it was the whole table that should be advertised, but that is perhaps in later version when we add versioning of the tables

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -539,72 +527,98 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
-  def encoder: Flow[Send, EnvelopeBuffer, NotUsed] = createEncoder(envelopePool)
+      val addr = localAddress.address
+      log.error(s"advertise compression to [$remoteAddress], of [$ref] => [$id] ")
```
debug (without allocations)

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala:
```diff
@@ -0,0 +1,138 @@
+   * @throws UnknownCompressedIdException if given id is not known, this may indicate a bug – such situation should not happen.
+   */
+  def decompress(idx: Int): T = {
+    log.debug(s"  --- decompress [$idx] => ${compressionAllocations.get(idx)}")
```
no allocations, please

ktoso replied:
done, and also hidden under additional compression.debug flag, it's only ever useful when debugging compression - not poluting general debug output

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala:
yes, we should, but not in this PR

ktoso commented:
rebase went wrong... w8 :(

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -206,8 +199,11 @@ private[akka] trait InboundContext {
+      incarnation = 1,
+      uniqueRemoteAddressPromise = Promise(),
+      quarantined = ImmutableLongMap.empty[QuarantinedTimestamp],
+      outboundCompression = NoOutboundCompression)
```
keeping outbound compression here makes sense, since it is per incarnation, and new incarnation means new AssociationState.

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -362,7 +362,11 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
+    s"aeron:udp?endpoint=${a.host.get}:${a.port.get}"
+  } catch {
+    case ex: Throwable ⇒ throw new RuntimeException("BOOM: " + a, ex)
+  }
```
removing this, not needed

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -384,10 +388,6 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
-  // FIXME: Compression table must be owned by each channel instead
-  // of having a global one
-  val compression = new Compression(system)
```
is now hosted in association state.
While it is true that we may be getting compression requests before the association is ready, this is OK, since then we say "outgoing compression not ready yet" (you'll see that in the encoder)

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -538,33 +538,37 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
   private def runInboundStreams(): Unit = {
-    runInboundControlStream()
-    runInboundOrdinaryMessagesStream()
+    val noCompression = new NoInboundCompression(system)
```
Adding TODO, about making it used in all channels (AFAIR thats what we agreed a bit back), as next step

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -538,33 +538,37 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
   }
 
-  private def runInboundControlStream(): Unit = {
+  private def runInboundControlStream(compression: InboundCompression): Unit = {
```
btw, the split between InboundCompression and OutboundCompression was a sanity life saver

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -135,17 +141,17 @@ private[remote] class Association(
           current.uniqueRemoteAddressValue() match {
             case Some(old) ⇒
-              log.debug(
+              log.info(
```
moved back to debug

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -396,7 +411,34 @@ private[remote] class Association(
+   * This proxy uses the current associationStates compression table, which is reset for a new incarnation.
+   * This way the same outgoing stream will switch to using the new table without the need of restarting it.
+   */
+  object CurrentAssociationStateOutboundCompressionProxy extends OutboundCompression {
```
This allows the active streams to access the "right" compression table, i.e. if we have a new associationState

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala:
```diff
@@ -68,19 +70,37 @@ private[remote] object EnvelopeBuffer {
+  def out(compression: OutboundCompression): HeaderBuilder = new HeaderBuilderImpl(NoopInboundCompression, compression)
+
+  /** INTERNAL API, FOR TESTING ONLY */
+  private[remote] def bothWays(in: InboundCompression, out: OutboundCompression): HeaderBuilder = new HeaderBuilderImpl(in, out)
```
Not happy about exposing it here, but saves us a lot of dancing around in `EnvelopeBufferSpec` (which both reads/writes the same one)

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/Codecs.scala:
```diff
@@ -32,58 +34,40 @@ private[remote] class Encoder(
+          //          // TODO hide under flag? I find these pretty interesting to inspect
+          //          import akka.util.PrettyByteString.asPretty
+          //          envelope.byteBuffer.flip()
+          //          log.info("OUT: " + ByteString(envelope.byteBuffer).prettyPrint(Int.MaxValue))
```
WDYT, can we keep it under hidden flag?

patriknw replied:
ok

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/compress/AdvertiseCompressionId.scala:
```diff
@@ -0,0 +1,11 @@
+import akka.actor.Address
+
+/** Callback invoked when a compression id allocation should be advertised to the remote actor system. */
+trait AdvertiseCompressionId[T] {
```
could be just a function, but much easier to track it when debugging given a proper name

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/compress/OutboundActorRefCompressionTable.scala:
```diff
@@ -0,0 +1,125 @@
+        if (settings.debug) log.debug("Outgoing: Updated compression table state: \n{}", toDebugString) // TODO debug  
+
+      case None ⇒
+        // TODO could be wrong? since we can not guarantee alocations come in sequence?
```
This TODO will be irrelevant very soon (entire compression advertised in one go)

ktoso commented on akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala:
```diff
@@ -0,0 +1,165 @@
+ * for a discussion about the assumptions made and guarantees about the Heavy Hitters made in this model.
+ * We assume the Cash Register model in which there are only additions, which simplifies HH detecion significantly.
+ */
+private[remote] final class TopHeavyHitters[T](val max: Int) {
```
This is the optimised version already

ktoso commented on akka-remote/src/test/scala/akka/remote/artery/compress/HeavyHittersSpec.scala:
```diff
@@ -0,0 +1,101 @@
+      hitters.update("A", 1) should ===(true)
+      hitters.update("B", 2) should ===(true)
+      hitters.update("C", 3) should ===(true) // A was dropped now  
+      hitters.update("A", 10) should ===(true) // TODO this is technically unexpected, we have already compressed A...  
```
This is a tradeoff, currently:

If someone was a HH, gets allocated an ID in the table, then stops being one and drops of our table, and then becomes one again (i.e. a few hitters competing for the lowest spot in the table), then we'll be trashing the table somewhat. We could avoid this by keeping a bloom filter of already allocated ids (in order to not keep EVERYTHING in memory), that increases mem footprint a bit.

Technically the scenario is not an issue esp. with entire table advertisement...

Yeah, not an issue, we could compress them later on before sending the table... Ignore this comment and this will be fixed with the 2 table mode :)

ktoso commented on project/plugins.sbt:
```diff
@@ -25,7 +25,9 @@ addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.3.3")
 addSbtPlugin("com.thoughtworks.sbt-api-mappings" % "sbt-api-mappings" % "0.2.2")
 
-addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.3")
+addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.6")
```
just a simple update

ktoso commented on project/plugins.sbt:
```diff
@@ -25,7 +25,9 @@ addSbtPlugin("com.eed3si9n" % "sbt-unidoc" % "0.3.3")
-addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.3")
+addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.6")
+
+addSbtPlugin("pl.project13.sbt" % "sbt-jol" % "0.1.1")
```
yay, jol!

patriknw replied:
where is it used here?

ktoso replied:
It is never used "in code", however it makes the `jol:...` commands available in the sbt console.

patriknw commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala:
```diff
@@ -64,7 +64,7 @@ private[http] final class PoolGateway(gatewayRef: ActorRef, val hcps: HostConnec
   override def toString = s"PoolGateway(hcps = $hcps)"
 
-  // INTERNAL API (testing only)
+  /** INTERNAL API (testing only) */
```
don't change unnecessary things since it may complicate merge back to master

ktoso replied:
:+1:

patriknw commented on akka-remote-tests/src/multi-jvm/scala/akka/remote/artery/MaxThroughputSpec.scala:
```diff
@@ -19,6 +18,7 @@ import akka.serialization.ByteBufferSerializer
 import akka.serialization.SerializerWithStringManifest
 import akka.testkit._
 import com.typesafe.config.ConfigFactory
+import java.net.InetAddress
```
?

patriknw commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,338 @@
+ * This implementation is mostly taken and adjusted from the Apache V2 licensed project `stream-lib`, located here:
+ * https://github.com/clearspring/stream-lib/blob/master/src/main/java/com/clearspring/analytics/stream/frequency/CountMinSketch.java
+ */
+public class CountMinSketch {
```
add internal api marker

patriknw commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,338 @@
+      if (o instanceof Float) {
+        return hashLong(Float.floatToRawIntBits((Float) o));
+      }
+      if (o instanceof String) {
```
move up, if `String` is what we mostly hash

ktoso replied:
good idea!

patriknw commented on akka-remote/src/main/java/akka/remote/artery/compress/CountMinSketch.java:
```diff
@@ -0,0 +1,338 @@
+      if (o instanceof byte[]) {
+        return hash((byte[]) o);
+      }
+      return hash(o.toString());
```
what do we hash? is it also `ActorRef`?
if that is the case we should use `ActorRef.hashCode`, it's super fast

ktoso replied:
Hm, great idea – yes we do use it on ActorRef, indirectly (passed in here it would have been a String, but we can change that).
Question is if that's going to work hm... `hashCode` seems to be `uid`, which is just a random number, CMS heavily relies on even hashcode distribution, thus murmuring keys.

I'm not sure the change would be valid, WDYT @drewhk ?

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -134,7 +134,37 @@ akka {
+          }
+          manifests {
+            # enables
+            enabled = off
```
when we have tried out things I think we can remove these `enabled` flags, I see no reason why one would not want to use compression

ktoso replied:
I'll add a todo

patriknw commented on akka-remote/src/main/resources/reference.conf:
```diff
@@ -134,7 +134,37 @@ akka {
+          manifests {
+            # enables
+            enabled = off
+            # Max number of compressed actor-refs
```
s/actor-refs/manifests/

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -247,14 +244,17 @@ private[akka] final class AssociationState(
+          incarnation,
+          uniqueRemoteAddressPromise,
+          quarantined = quarantined.updated(a.uid, QuarantinedTimestamp(System.nanoTime())),
+          outboundCompression = outboundCompression)
```
after quarantine it can be cleared with `NoOutboundCompression`

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -247,14 +244,17 @@ private[akka] final class AssociationState(
-  def newIncarnation(remoteAddressPromise: Promise[UniqueAddress]): AssociationState =
-    new AssociationState(incarnation + 1, remoteAddressPromise, quarantined)
+  def newIncarnation(remoteAddressPromise: Promise[UniqueAddress], compression: OutboundCompression): AssociationState =
+    new AssociationState(incarnation + 1, remoteAddressPromise, quarantined, compression)
```
shouldn't the table be cleared here? this is called when handshake is completed

patriknw replied:
scratch that, I see that it's cleared

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala:
```diff
@@ -728,39 +742,46 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
+    else new NoInboundCompression(system)
+
+  def createEncoder(pool: EnvelopeBufferPool, compression: OutboundCompression): Flow[Send, EnvelopeBuffer, NotUsed] =
+    Flow.fromGraph(new Encoder(localAddress, system, compression, pool)) // TODO pass in
```
obsolete TODO?

ktoso replied:
yes, thanks

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -79,6 +81,10 @@ private[remote] class Association(
 
+  def compression: OutboundCompression = associationState.outboundCompression
+  //    _compression
+  //  @volatile private[this] var _compression: OutboundCompression = _ // will be initialized in association()
```
?

ktoso replied:
old, removed

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -326,53 +341,53 @@ private[remote] class Association(
       } else {
         Source.fromGraph(new SendQueue[Send])
-          .toMat(transport.outbound(this))(Keep.both)
+          .toMat(transport.outbound(this, compression))(Keep.both) // ????????
```
?

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -396,7 +411,34 @@ private[remote] class Association(
   }
 
-  override def toString(): String =
+  // FIXME: Compression table must be owned by each channel instead of having a global one
```
obsolete FIXME?

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Association.scala:
```diff
@@ -396,7 +411,34 @@ private[remote] class Association(
+  private def createOutboundCompressionTable(remoteAddress: Address): OutboundCompression = {
+    if (transport.provider.remoteSettings.ArteryCompressionSettings.enabled) {
+      val compression = new OutboundCompressionImpl(transport.system, remoteAddress)
+      log.info("Creating Outbound compression table ({}) to [{}]", compression.hashCode, remoteAddress)
```
add a FIXME that it should use a version number instead of the `hashCode`

ktoso replied:
yup, thanks

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala:
```diff
@@ -4,13 +4,15 @@
-import sun.misc.Cleaner
-import akka.util.Unsafe
+import akka.util.{ OptionVal, Unsafe }
+import sun.management.snmp.jvminstr.JVM_MANAGEMENT_MIB_IMPL
```
?

ktoso replied:
Whoa, no idea how that got imported; removed

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala:
```diff
@@ -129,64 +151,61 @@ private[remote] final class HeaderBuilderImpl(val compressionTable: LiteralCompr
-    _recipientActorRefIdx = compressionTable.compressActorRef(ref)
+  def recipientActorRef_=(ref: ActorRef): Unit = {
+    _recipientActorRefIdx = outboundCompression.compressActorRef(ref)
+    if (_recipientActorRefIdx == -1) _recipientActorRef = ref.path.toSerializationFormat
```
I wonder what we should do with `ref.path.toSerializationFormat`. That is super slow and I think it might be worth caching it. Perhaps we have a cached field for it in `RemoteActorRef`?

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Codecs.scala:
```diff
@@ -156,35 +137,32 @@ private[remote] class Decoder(
+
+        val sender: InternalActorRef = headerBuilder.senderActorRef match {
+          case OptionVal.Some(ref) ⇒ ref.asInstanceOf[InternalActorRef]
+          case _                   ⇒ resolveActorRefWithLocalAddress(headerBuilder.senderActorRefPath)
```
why `headerBuilder.senderActorRefPath` and not `headerBuilder.senderActorRef`?

ktoso replied:
in the `case _ =>` the value of `senderActorRef` is `OptionVal.None`, i.e. it was not compressed.
Compression acts like a cache here, in the sense of that we directly get the actual ActorRef back.
If we don't, we need to get the "raw string" and resolve it.

Not sure if we'd rather mvoe this into the header builder... it would need more context AFAICS (what local address is)

patriknw replied:
ok, I see now.
`OptionVal.None` might be easier to read.
We might need some caching of the resolved actor refs also, in addition to what the compression table holds. resolve is super slow and we want to return the same `RemoteActorRef` instances as much as possible. Not sure what is best. Similar for recipient.

This is ok for this PR.

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Codecs.scala:
```diff
@@ -218,28 +196,11 @@ private[remote] class Decoder(
+        resolveActorRefWithLocalAddress(path) match {
+          case empty: EmptyLocalActorRef ⇒
+            val pathElements = empty.path.elements
+            if (pathElements.nonEmpty && pathElements.head == "remote") OptionVal.None
```
This is for the remote deployment corner case, we need to do this only once, but for this PR it can be like this. I can fix that when it has been merged. Please add a FIXME.

ktoso replied:
added fixme

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/Codecs.scala:
```diff
@@ -254,11 +215,10 @@ private[remote] class Decoder(
                   val recipient = resolveActorRefWithLocalAddress(recipientPath)
-                  // only retry for the first message
-                  recipientCache.put(recipientPath, recipient)
+                  // FIXME only retry for the first message, need to keep them in a cache
```
ah, here is the FIXME I mentioned above

ktoso replied:
ah, okey

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/NoLiteralCompression.scala:
```diff
@@ -0,0 +1,38 @@
+final class NoInboundCompression(system: ActorSystem) extends InboundCompression {
+  override def hitActorRef(address: Address, ref: ActorRef): Unit = ()
+  override def decompressActorRef(idx: Int): OptionVal[ActorRef] =
+    if (idx == -1) throw new scala.IllegalArgumentException("Attemted decompression of illegal compression id: -1")
```
remove `scala.`

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/ActualCompressionTables.scala:
```diff
@@ -0,0 +1,111 @@
+  // class manifest compression --- 
+
+  override def compressClassManifest(manifest: String): Int = {
+    val id = classManifestsOut.compress(manifest)
```
no need for `val`

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/CompressionSettings.scala:
```diff
@@ -0,0 +1,43 @@
+    val max = c.getInt("max")
+  }
+
+  object serializers {
```
remove

patriknw commented on akka-remote/src/main/scala/akka/remote/artery/compress/InboundCompressionTable.scala:
```diff
@@ -0,0 +1,161 @@
+ * [3]> the callback for example then triggers an CompressionAdvertisement to the receiving side
+ */
+// TODO should the onHeavyHitter be inside HeavyHitters?
+class InboundCompressionTable[T](
```
`private[remote]` and internal api marker, here and elsewehere

patriknw commented on akka-stream/src/main/scala/akka/stream/impl/ConstantFun.scala:
```diff
@@ -34,4 +35,6 @@ private[akka] object ConstantFun {
 
   val two2none = (_: Any, _: Any) ⇒ None
+
+  val two2unit = (_: Any, _: Any) ⇒ ()
```
why these changes?

ktoso replied:
Good catch, not used anymore (they were at some point, instead of explicit NoOp objects).
Removing

patriknw commented on akka-testkit/src/main/scala/akka/testkit/TestActors.scala:
```diff
@@ -20,6 +20,15 @@ object TestActors {
   /**
+   * BlackholeActor does nothing for incoming messages, its like a blackhole.
+   */
+  class BlackholeActor extends Actor {
```
not used in this PR

patriknw replied:
wrong search, ignore that

ktoso replied:
Yup, that one is used as I only wanted "one direction" triggered compression

patriknw commented:
LGTM, after minor things

ktoso commented on akka-remote/src/test/scala/akka/remote/artery/compress/OutboundCompressionTableSpec.scala:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    // TODO cover more cases of holes in the redeliveries of advertisements
+    // TODO ^ to cover the fast forward logic a bit more
```
fast-forward logic is not needed to be covered much since we're going to the advertise entire table logic

ktoso commented:
Pushed follow up to all comments. Should pass validation now as well (failed on compiling JMH benchmark previously, which I had forgotten about)

ktoso commented:
All green and now onwards to moving to the other table advertisement mode
